File: acwing/算法基础课/第四讲/883.高斯消元解线性方程组.py
N = 110
arr = [[0 for _ in range(N)]for _ in range(N)]
eps = 1e-6


def gauss():
    r = 0

    for c in range(n):
        t = r
        for i in range(r, n):
            if abs(arr[i][c]) > abs(arr[t][c]):
                t = i

        if abs(arr[t][c]) < eps:
            continue

        # swap(arr[c])
        # for i in range(c, n + 1):
        #     arr[t][i], arr[r][i] = arr[r][i], arr[t][i]
        arr[t], arr[r] = arr[r], arr[t] 
        for i in range(n, c - 1, -1):
            arr[r][i] /= arr[r][c]
        for i in range(r + 1, n):
            if abs(arr[i][c]) > eps:
                for j in range(n, c - 1, -1):
                    arr[i][j] -= arr[i][c] * arr[r][j]
        r += 1

    if r < n:
        for i in range(r, n):
            if abs(arr[i][n]) > eps:
                return 0
        return 2
    for i in range(n - 1, -1, -1):
        for j in range(i + 1, n):
            arr[i][n] -= arr[i][j] * arr[j][n]
    return 1


n = int(input())

for i in range(n):
    arr[i] = list(map(float, input().split()))

t = gauss()

if t == 0:
    print("No solution")
elif t == 2:
    print("Infinite group solutions")
elif t == 1:
    for i in range(n):
        print("{:.2f}".format(arr[i][n]))

File: acwing/算法基础课/第四讲/883.高斯消元解线性方程组/883.高斯消元解线性方程组.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <cstdio>
#include <cmath>

using namespace std;

const int N = 105;
const double eps = 1e-8;

double a[N][N];

int n;

int gauss()
{
    int c, r;
    for (c = 0, r = 0; c < n; c++)
    {
        int t = r;
        for (int i = r; i < n; i++)
        {
            if (fabs(a[i][c]) > fabs(a[t][c]))
                t = i;
        }
        if (fabs(a[t][c]) < eps)
            continue;

        for (int i = c; i <= n; i++)
            swap(a[r][i], a[t][i]);
        for (int i = n; i >= c; i--)
            a[r][i] /= a[r][c];
        for (int i = r + 1; i < n; i++)
        {
            if (fabs(a[i][c]) > eps)
            {
                for (int j = n; j >= c; j--)
                {
                    a[i][j] -= a[i][c] * a[r][j];
                }
            }
        }
        r++;
    }
    if (r < n)
    {
        for (int i = r; i < n; i++)
        {
            if (fabs(a[i][n]) > eps)
                return 2;
        }
        return 1;
    }

    for (int i = n - 1; i >= 0; i--)
    {
        for (int j = i + 1; j < n; j++)
        {
            a[i][n] -= a[j][n] * a[i][j];
        }
    }

    return 0;
}

int main()
{
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j <= n; j++)
        {
            cin >> a[i][j];
        }
    }
    int t = gauss();
    if (t == 1)
    {
        cout << "Infinite group solutions" << endl;
    }
    else if (t == 2)
    {
        cout << "No solution" << endl;
    }
    else
    {
        for (int i = 0; i < n; i++)
        {
            printf("%.2lf\n", a[i][n]);
        }
    }
    return 0;
}